File: convert_csv_to_geojson.js
/**
 * Convert merged till 26.csv to GeoJSON format for ManholeCovers_1.js
 */

const fs = require('fs');
const path = require('path');

function parseCSV(csvContent) {
    const lines = csvContent.split('\n');
    
    // Skip first 3 comment lines
    const dataLines = lines.slice(3);
    
    // Parse header
    const header = dataLines[0].split(',').map(h => h.trim());
    
    const features = [];
    const skippedRows = [];
    let fid = 0;
    let srNo = 1;
    
    for (let i = 1; i < dataLines.length; i++) {
        const line = dataLines[i].trim();
        if (!line) continue;
        
        // Simple CSV parsing (handles quoted fields)
        const values = [];
        let current = '';
        let inQuotes = false;
        
        for (let char of line) {
            if (char === '"') {
                inQuotes = !inQuotes;
            } else if (char === ',' && !inQuotes) {
                values.push(current.trim());
                current = '';
            } else {
                current += char;
            }
        }
        values.push(current.trim());
        
        // Create object from header and values
        const row = {};
        header.forEach((h, idx) => {
            row[h] = values[idx] || '';
        });
        
        // Skip rows with invalid coordinates
        const lat = parseFloat(row['Latitude']);
        const lon = parseFloat(row['Longitude']);
        
        if (isNaN(lat) || isNaN(lon) || row['Latitude'] === 'Not found' || row['Longitude'] === 'Not found') {
            console.log(`Skipping row ${i} with invalid coordinates`);
            skippedRows.push({
                rowNumber: i,
                reason: isNaN(lat) ? 'Invalid Latitude' : isNaN(lon) ? 'Invalid Longitude' : 'Not found',
                ...row
            });
            continue;
        }
        
        // Extract fields
        const location = row['Location/Plus Code'].replace(/"/g, '');
        const timestamp = row['Timestamp'] || 'Not found';
        const pics = row['PICS'] || '';
        
        // Create feature
        const feature = {
            "type": "Feature",
            "properties": {
                "FID": String(fid),
                "Shape *": "Point",
                "Sr_No": String(srNo),
                "Latitude": lat,
                "Longitude": lon,
                "Location": location,
                "ZONE": "A",
                "COORDINATE": `${lat} , ${lon}`,
                "PICS": `images\\${pics}`
            },
            "geometry": {
                "type": "Point",
                "coordinates": [lon, lat]
            }
        };
        
        features.push(feature);
        fid++;
        srNo++;
    }
    
    return { features, skippedRows };
}

function exportSkippedRowsToCSV(skippedRows, outputFile) {
    if (skippedRows.length === 0) {
        console.log('No skipped rows to export.');
        return;
    }
    
    // Get all unique keys from skipped rows
    const allKeys = new Set();
    skippedRows.forEach(row => {
        Object.keys(row).forEach(key => allKeys.add(key));
    });
    
    const headers = ['Row Number', 'Reason', ...Array.from(allKeys).filter(k => k !== 'rowNumber' && k !== 'reason')];
    
    // Create CSV content
    let csvContent = headers.join(',') + '\n';
    
    skippedRows.forEach(row => {
        const values = headers.map(header => {
            let value = '';
            if (header === 'Row Number') value = row.rowNumber;
            else if (header === 'Reason') value = row.reason;
            else value = row[header] || '';
            
            // Escape quotes and wrap in quotes if contains comma
            if (typeof value === 'string' && (value.includes(',') || value.includes('"') || value.includes('\n'))) {
                value = '"' + value.replace(/"/g, '""') + '"';
            }
            return value;
        });
        csvContent += values.join(',') + '\n';
    });
    
    fs.writeFileSync(outputFile, csvContent, 'utf-8');
    console.log(`\nExported ${skippedRows.length} skipped rows to: ${outputFile}`);
}

function main() {
    const csvFile = path.join(__dirname, 'data', 'merged till 26.csv');
    const outputFile = path.join(__dirname, 'data', 'ManholeCovers_1.js');
    const skippedFile = path.join(__dirname, 'data', 'skipped_rows.csv');
    
    // Read CSV file
    console.log(`Reading CSV file: ${csvFile}`);
    const csvContent = fs.readFileSync(csvFile, 'utf-8');
    
    // Parse CSV
    console.log('Parsing CSV...');
    const { features, skippedRows } = parseCSV(csvContent);
    
    // Create GeoJSON structure
    const geojson = {
        "type": "FeatureCollection",
        "name": "ManholeCovers_1",
        "crs": {
            "type": "name",
            "properties": {
                "name": "urn:ogc:def:crs:OGC:1.3:CRS84"
            }
        },
        "features": features
    };
    
    // Write to JavaScript file
    console.log(`Writing output to: ${outputFile}`);
    const jsContent = 'var json_ManholeCovers_1 = ' + JSON.stringify(geojson, null, 2) + ';';
    fs.writeFileSync(outputFile, jsContent, 'utf-8');
    
    console.log(`\nConversion complete!`);
    console.log(`Created ${features.length} features.`);
    console.log(`Output written to: ${outputFile}`);
    
    // Export skipped rows
    exportSkippedRowsToCSV(skippedRows, skippedFile);
}

main();

File: convert_csv_to_geojson.py
#!/usr/bin/env python3
"""
Convert merged till 26.csv to GeoJSON format for ManholeCovers_1.js
"""

import csv
import json
import os

def convert_csv_to_geojson(csv_file, output_file):
    """Convert CSV file to GeoJSON format matching ManholeCovers_1.js structure"""
    
    features = []
    fid_counter = 0
    sr_no_counter = 1
    
    with open(csv_file, 'r', encoding='utf-8') as f:
        # Skip the first 3 comment lines
        for _ in range(3):
            next(f)
        
        reader = csv.DictReader(f)
        
        for row in reader:
            # Skip rows with missing coordinates
            try:
                lat = float(row['Latitude'])
                lon = float(row['Longitude'])
            except (ValueError, KeyError):
                print(f"Skipping row with invalid coordinates: {row}")
                continue
            
            # Handle "Not found" values
            if row['Latitude'] == 'Not found' or row['Longitude'] == 'Not found':
                print(f"Skipping row with 'Not found' coordinates: {row}")
                continue
            
            # Extract location from Location/Plus Code field
            location = row.get('Location/Plus Code', '').strip('"')
            
            # Extract timestamp
            timestamp = row.get('Timestamp', 'Not found')
            
            # Extract image filename
            pics = row.get('PICS', '')
            
            # Create feature
            feature = {
                "type": "Feature",
                "properties": {
                    "FID": str(fid_counter),
                    "Shape *": "Point",
                    "Sr_No": str(sr_no_counter),
                    "Latitude": lat,
                    "Longitude": lon,
                    "Location": location,
                    "ZONE": "A",  # Default zone
                    "COORDINATE": f"{lat} , {lon}",
                    "PICS": f"images\\{pics}"  # Assuming images are in images folder
                },
                "geometry": {
                    "type": "Point",
                    "coordinates": [lon, lat]
                }
            }
            
            features.append(feature)
            fid_counter += 1
            sr_no_counter += 1
    
    # Create GeoJSON structure
    geojson = {
        "type": "FeatureCollection",
        "name": "ManholeCovers_1",
        "crs": {
            "type": "name",
            "properties": {
                "name": "urn:ogc:def:crs:OGC:1.3:CRS84"
            }
        },
        "features": features
    }
    
    # Write to JavaScript file
    with open(output_file, 'w', encoding='utf-8') as f:
        f.write('var json_ManholeCovers_1 = ')
        json.dump(geojson, f, indent=2, ensure_ascii=False)
        f.write(';')
    
    print(f"Conversion complete! Created {len(features)} features.")
    print(f"Output written to: {output_file}")

if __name__ == "__main__":
    csv_file = "data/merged till 26.csv"
    output_file = "data/ManholeCovers_1.js"
    
    if not os.path.exists(csv_file):
        print(f"Error: CSV file not found: {csv_file}")
        exit(1)
    
    convert_csv_to_geojson(csv_file, output_file)
